libvast/src/arrow_table_slice.cpp

```diff
@@ -986,196 +789,72 @@ auto record_at(const record_type& type, const arrow::StructArray& arr,
-      return;
-    result_ = list_at(t.value_type(), arr, row_);
-  }
-
-  void operator()(const arrow::MapArray& arr, const map_type& t) {
-    if (arr.IsNull(row_))
-      return;
-    result_ = map_at(t.key_type(), t.value_type(), arr, row_);
-  }
-
-  void operator()(const arrow::StructArray& arr, const record_type& t) {
-    if (arr.IsNull(row_))
-      return;
-    result_ = record_at(t, arr, row_);
-  }
-
-private:
-  data_view result_;
-  int64_t row_;
-};
-
-template <class Array>
-data_view value_at(const type& t, const Array& arr, size_t row) {
-  row_picker f{row};
-  decode(t, arr, f);
-  return std::move(f.result());
+data_view value_at(const type& t, const arrow::Array& arr, int64_t row) {
+  auto f = detail::overload{
+    [&](const bool_type&, const arrow::BooleanArray& a) {
+      return a.Value(row);
+    },
+    [&](const integer_type&, const arrow::Int64Array& a) {
+      return integer{a.Value(row)};
+    },
+    [&](const count_type&, const arrow::UInt64Array& a) {
+      return a.Value(row);
+    },
+    [&](const real_type&, const arrow::DoubleArray& a) {
+      return a.Value(row);
+    },
+    [&](const enumeration_type&, const enum_array& a) {
+      return enumeration_at(
+        static_cast<const arrow::DictionaryArray&>(*a.storage()), row);
+    },
+    [&](const time_type&, const arrow::TimestampArray& a) {
+      return timestamp_at(a, row);
+    },
+    [&](const duration_type&, const arrow::DurationArray& a) {
+      return duration_at(a, row);
+    },
+    [&](const address_type&, const address_array& a) {
+      return address_at(
+        static_cast<const arrow::FixedSizeBinaryArray&>(*a.storage()), row);
+    },
+    [&](const subnet_type&, const subnet_array& a) {
+      return subnet_at(static_cast<const arrow::StructArray&>(*a.storage()),
+                       row);
+    },
+    [&](const string_type&, const arrow::StringArray& a) {
+      return string_at(a, row);
+    },
+    [&](const pattern_type&, const pattern_array& a) {
+      return pattern_view{
+        string_at(static_cast<const arrow::StringArray&>(*a.storage()), row)};
+    },
+    [&](const list_type& lt, const arrow::ListArray& a) {
+      return list_at(lt.value_type(), a, row);
+    },
+    [&](const map_type& mt, const arrow::MapArray& a) {
+      return map_at(mt.key_type(), mt.value_type(), a, row);
+    },
+    [&](const record_type& rt, const arrow::StructArray& a) {
+      return record_at(rt, a, row);
+    },
+    [&](const auto&, const auto&) -> data_view {
+      die(fmt::format("unhandled type pair({}, {})!\n", t,
+                      arr.type()->ToString()));
+    },
+  };
+  if (arr.IsNull(row))
+    return caf::none;
+  return caf::visit(f, t, arr);
 }
 
-class index_applier {
-public:
-  using result_type = void;
-
-  index_applier(size_t offset, value_index& idx)
-    : offset_(detail::narrow_cast<int64_t>(offset)), idx_(idx) {
-    // nop
+auto apply_index(size_t offset, value_index& idx, const arrow::Array& arr,
```

Comment:
I imagine that we don't need this function at all. Instead, let's add something like this:

```cpp
auto values(size_t column, const concrete_type auto& type, const auto& array)
    -> detail::generator<view<type_to_data_t<decltype(type)>>> {
  ...
}
```

And then a second version that is generic:

```cpp
auto values(size_t column, const type& type, const arrow::Array& array)
    -> detail::generator<data_view> {
  auto f = [column](const concrete_type auto& type, const auto& array) {
    return values(column, type, array);
  };
  for (auto&& result : caf::visit(f, type, array))
    co_yield result;
}
```

After this, we can simply use this in `apply_column_to_index`:

```cpp
if (const auto& batch = record_batch()) {
  const auto& layout = caf::get<record_type>(this->layout());
  auto type = layout.field(layout.resolve_flat_index(column)).type;
  const auto& array = state_.flat_columns[column];
  for (size_t row = 0; auto&& value : values(column, type, array)) {
    if (!caf::holds_alternative<view<caf::none_t>>(value))
      idx.append(value, offset + row);
    ++row;
  }
}